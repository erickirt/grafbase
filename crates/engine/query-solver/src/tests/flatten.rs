use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    A @join__graph(name: "a", url: "http://localhost:4200/requires-with-argument/a")
    B @join__graph(name: "b", url: "http://localhost:4200/requires-with-argument/b")
    C @join__graph(name: "c", url: "http://localhost:4200/requires-with-argument/c")
    D @join__graph(name: "d", url: "http://localhost:4200/requires-with-argument/d")
}

type Author implements Node
    @join__type(graph: D)
{
    id: ID!
    name: String
}

type Comment implements Node
    @join__type(graph: C, key: "id")
    @join__type(graph: D, key: "id")
{
    id: ID!
    authorId: ID @join__field(graph: C)
    body: String! @join__field(graph: C)
    date: String @join__field(graph: D)
}

type Post implements Node
    @join__type(graph: C, key: "id")
    @join__type(graph: D, key: "id")
{
    id: ID!
    author: Author @join__field(graph: D, requires: "comments(limit: 3) { authorId }")
    comments(limit: Int!): [Comment] @join__field(graph: D)
}

type Product
    @join__type(graph: A, key: "upc")
    @join__type(graph: B, key: "upc")
{
    upc: String!
    weight: Int @join__field(graph: B)
    price(currency: String!): Int @join__field(graph: B)
    shippingEstimate: Int @join__field(graph: A, requires: "price(currency: \"USD\") weight")
    name: String @join__field(graph: B)
}

interface Node {
    id: ID!
}

type Query
    @join__type(graph: A)
    @join__type(graph: B)
    @join__type(graph: C)
    @join__type(graph: D)
{
    node: Node @join__field(graph: D)
    products: [Product] @join__field(graph: B)
    feed: [Post] @join__field(graph: C)
}
"#;

#[test]
fn comments_fields_should_be_flattened() {
    assert_solving_snapshots!(
        "comments_fields_should_be_flattened",
        SCHEMA,
        r#"
        query ($limit: Int = 1) {
          feed {
            author {
              id
            }
            ...Foo
            ...Bar
          }
        }

        fragment Foo on Post {
          comments(limit: $limit) {
            id
          }
        }

        fragment Bar on Post {
          comments(limit: $limit) {
            id
          }
        }
        "#
    );
}

#[test]
fn flattening_cannot_merge_fields_with_different_type_conditions() {
    assert_solving_snapshots!(
        "flattening_cannot_merge_fields_with_different_type_conditions",
        SCHEMA,
        r#"
        query {
          node {
            ... on Post {
              ... on Node {
                id
              }
            }
            ... on Comment {
              ... on Node {
                id
              }
            }
          }
        }
        "#
    );
}
