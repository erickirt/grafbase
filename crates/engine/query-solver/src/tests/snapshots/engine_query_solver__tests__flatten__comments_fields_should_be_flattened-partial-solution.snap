---
source: crates/engine/query-solver/src/tests/flatten.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Root#c\", color=royalblue,shape=parallelogram ]\n    2 [ label = \"Query.feed\" ]\n    3 [ label = \"FedEntity#d\", color=royalblue,shape=parallelogram ]\n    4 [ label = \"Post.author\" ]\n    5 [ label = \"Author.id\" ]\n    6 [ label = \"Post.comments\" ]\n    7 [ label = \"Comment.id\" ]\n    8 [ label = \"*Post.comments\" ]\n    9 [ label = \"FedEntity#c\", color=royalblue,shape=parallelogram ]\n    10 [ label = \"*Comment.authorId\" ]\n    11 [ label = \"*Comment.id\" ]\n    12 [ label = \"*Post.id\" ]\n    0 -> 1 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 2 [  ]\n    2 -> 3 [ color=royalblue,fontcolor=royalblue ]\n    3 -> 4 [  ]\n    4 -> 5 [  ]\n    3 -> 6 [  ]\n    6 -> 7 [  ]\n    3 -> 8 [  ]\n    8 -> 9 [ color=royalblue,fontcolor=royalblue ]\n    9 -> 10 [  ]\n    8 -> 11 [  ]\n    2 -> 12 [  ]\n    3 -> 12 [ color=orangered,arrowhead=inv ]\n    4 -> 10 [ color=orangered,arrowhead=inv ]\n    4 -> 8 [ color=orangered,arrowhead=inv ]\n    9 -> 11 [ color=orangered,arrowhead=inv ]\n}\n"
---
digraph {
    0 [ label = "root" ]
    1 [ label = "Root#c" ]
    2 [ label = "Query.feed" ]
    3 [ label = "FedEntity#d" ]
    4 [ label = "Post.author" ]
    5 [ label = "Author.id" ]
    6 [ label = "Post.comments" ]
    7 [ label = "Comment.id" ]
    8 [ label = "*Post.comments" ]
    9 [ label = "FedEntity#c" ]
    10 [ label = "*Comment.authorId" ]
    11 [ label = "*Comment.id" ]
    12 [ label = "*Post.id" ]
    0 -> 1 [ label = "QueryPartition" ]
    1 -> 2 [ label = "Field" ]
    2 -> 3 [ label = "QueryPartition" ]
    3 -> 4 [ label = "Field" ]
    4 -> 5 [ label = "Field" ]
    3 -> 6 [ label = "Field" ]
    6 -> 7 [ label = "Field" ]
    3 -> 8 [ label = "Field" ]
    8 -> 9 [ label = "QueryPartition" ]
    9 -> 10 [ label = "Field" ]
    8 -> 11 [ label = "Field" ]
    2 -> 12 [ label = "Field" ]
    3 -> 12 [ label = "RequiredBySubgraph" ]
    4 -> 10 [ label = "RequiredBySubgraph" ]
    4 -> 8 [ label = "RequiredBySubgraph" ]
    9 -> 11 [ label = "RequiredBySubgraph" ]
}
